--with-netmap does not compile

When configue--with-netmap:
../elements/userlevel/fromnetmapdevice.cc:189:1: error: no declaration matches ‘PacketBatch* FromNetmapDevice::pull_batch(int, unsigned int)’
 FromNetmapDevice::pull_batch(int port, unsigned max) {

If HAVE_NETMAP_PACKET_POOL is not set, pull_batch() is declared as a noop in fromnetmapdevice.cc line 189
So it eems to me that --enable-netmap-pool is the only viable solution to get batch processing

--with-netmap --enable-netmap-pool
../include/click/packet.hh:3153:21: error: ‘NetmapBufQ’ has not been declared
                 if (NetmapBufQ::is_valid_netmap_buffer(head))

In packet.hh, there is a reference to NetmapBufQ, declared in include/click/netmapdevice.hh. On one hand, one might say that packet.hh should not depend on elements, on the other hand, it might be fair enough if the user specifies "--with-netmap".

Unfortunately, netmapdevice.hh also includes packet.hh, so packet.hh cannot just include netmapdevice.hh.

I suggest that NetmapBufQ is moved to a header file of its own, eg.: netmapbufq.hh, and the interfaces: is_netmap_packet() and is_valid_netmap_packet() are changed to take pointers (char* and buffer_destructor*) instead pf Packet*, so that netmapbufq.hh does not depend on packet.hh

Finally, there seems to be something with NetmapDevice::make_netmap(): p->initialize() should take an arg, and WritablePacket::make_local_packet_pool() does not seem to exist (probably not to be used, since initialize_local_packet_pool() has been called previouslu by the thread itself. I suppose, local_packet_pool() is to be called instead. but that only works if HAVE_MULTITHREAD is set.

So, the mandatory config for netmap is --with-netmap --enable-netmap-pool --enable-user-multithread as I can see it.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -20,7 +20,7 @@ jobs:
     strategy:
         matrix:
             COMPILER: [gcc, clang]
-            FRAMEWORK: [vanilla, dpdk, dpdk-light, vanilla-nopool]
+            FRAMEWORK: [vanilla, dpdk, dpdk-light, netmap, vanilla-nopool]
 
     steps:
     - uses: actions/checkout@v2
@@ -55,6 +55,13 @@ jobs:
 
             if [ "${{matrix.FRAMEWORK}}" = "dpdk-light" ] ; then
                 FRAMEWORK_FLAGS="$FRAMEWORK_FLAGS --enable-dpdk-packet --disable-clone --disable-task-stats --disable-cpu-load --disable-dpdk-softqueue"
+            fi
+            if [ "${{matrix.FRAMEWORK}}" = "netmap" ] ; then
+                if [ ! -e "netmap-13.0" ] ; then
+                   wget https://github.com/luigirizzo/netmap/archive/refs/tags/v13.0.tar.gz
+                   tar -zxf v13.0.tar.gz
+                fi
+                FRAMEWORK_FLAGS="--with-netmap=$(pwd)/netmap-13.0/sys/"
             fi
             echo "FRAMEWORK_FLAGS=$FRAMEWORK_FLAGS" >> $GITHUB_ENV
             echo "PKG_CONFIG_PATH=$PKG_CONFIG_PATH" >> $GITHUB_ENV

--- a/elements/userlevel/fromnetmapdevice.cc
+++ b/elements/userlevel/fromnetmapdevice.cc
@@ -26,6 +26,7 @@
 #include <click/standard/scheduleinfo.hh>
 #include <click/packet.hh>
 #include <click/packet_anno.hh>
+#include <click/netmapdevice.hh>
 #include <vector>
 
 CLICK_DECLS
@@ -183,60 +184,6 @@ FromNetmapDevice::initialize(ErrorHandler *errh)
 
     return 0;
 }
-
-#if !HAVE_NETMAP_PACKET_POOL
-PacketBatch*
-FromNetmapDevice::pull_batch(int port, unsigned max) {
-		click_chatter("PULL BATCH only supports netmap batch!");
-		return 0;
-}
-#else
-PacketBatch*
-FromNetmapDevice::pull_batch(int port, unsigned max) {
-
-	for (int i = queue_for_thisthread_begin(); i <= queue_for_thisthread_end(); i++) {
-		lock();
-		struct nm_desc* nmd = _device->nmds[i];
-#ifdef SYNC_ON_IRQ
-		ioctl(nmd->fd,NIOCRXSYNC,SYNC_ON_IRQ);
-#else
-		ioctl(nmd->fd,NIOCRXSYNC,0);
-#endif
-		struct netmap_ring *rxring = NETMAP_RXRING(nmd->nifp, i);
-
-		u_int cur, n;
-
-		cur = rxring->cur;
-
-		n = nm_ring_space(rxring);
-		if (max && n > max) {
-			n = max;
-		}
-
-		if (n == 0) {
-			unlock();
-			continue;
-		}
-
-		Timestamp ts = Timestamp::make_usec(nmd->hdr.ts.tv_sec, nmd->hdr.ts.tv_usec);
-
-		PacketBatch *batch_head = WritablePacket::make_netmap_batch(n,rxring,cur,_set_rss_aggregate);
-		if (!batch_head) goto error;
-		batch_head->set_timestamp_anno(ts);
-		rxring->head = rxring->cur = cur;
-		unlock();
-		if (batch_head) {
-			add_count(batch_head->count());
-			return batch_head;
-		}
-	}
-	return 0;
-	error: //No more buffer
-	click_chatter("No more buffers !");
-	router()->master()->kill_router(router());
-	return 0;
-}
-#endif
 
 inline bool
 FromNetmapDevice::receive_packets(Task* task, int begin, int end, bool fromtask)

--- a/elements/userlevel/fromnetmapdevice.hh
+++ b/elements/userlevel/fromnetmapdevice.hh
@@ -4,7 +4,6 @@
 #include <click/config.h>
 #include <click/task.hh>
 #include <click/etheraddress.hh>
-#include <click/netmapdevice.hh>
 #include "queuedevice.hh"
 #include <vector>
 #include <stdint.h>
@@ -14,6 +13,8 @@
 #include <fstream>
 
 CLICK_DECLS
+
+class NetmapDevice;
 
 /*
  * =c

--- a/elements/userlevel/netmapinfo.cc
+++ b/elements/userlevel/netmapinfo.cc
@@ -18,6 +18,7 @@
 #include <click/config.h>
 #include <click/args.hh>
 #include <click/error.hh>
+#include <click/netmapdevice.hh>
 #include "netmapinfo.hh"
 
 CLICK_DECLS

--- a/elements/userlevel/netmapinfo.hh
+++ b/elements/userlevel/netmapinfo.hh
@@ -2,7 +2,6 @@
 #define CLICK_NETMAPINFO_HH
 
 #include <click/element.hh>
-#include <click/netmapdevice.hh>
 
 CLICK_DECLS
 

--- a/elements/userlevel/queuedevice.hh
+++ b/elements/userlevel/queuedevice.hh
@@ -204,6 +204,9 @@ protected:
         return queue_for_thread_end(click_current_cpu_id());
     }
 
+    /**
+     * Return the queue index for a given thread
+     */
     inline int id_for_thread(int tid) {
         if (likely(queue_per_threads == 1))
             return queue_for_thread_begin(tid) - firstqueue;
@@ -211,22 +214,37 @@ protected:
             return (queue_for_thread_begin(tid) - firstqueue) / queue_per_threads;
     }
 
+    /**
+     * Returns the queue id for this thread
+     */
     inline int id_for_thread() {
         return id_for_thread(click_current_cpu_id());
     }
 
+    /**
+     * Return the task responsible for the current thread
+     */
     inline Task* task_for_thread() {
         return _thread_state->task;
     }
 
+    /**
+     * Returns the task for a thread
+     */
     inline Task* task_for_thread(int tid) {
         return _thread_state.get_value_for_thread(tid).task;
     }
 
+    /**
+     * Returns the thread managing a given queue, given by its index (not accountig for first_queue, hence the queue offset)
+     */
     inline int thread_for_queue_offset(int queue) {
         return _q_infos[queue].thread_id;
     }
 
+    /**
+     * Returns the number of threads per queues, in general 1 except when the number of queues is limited
+     */
     int thread_per_queues() {
         return queue_share;
     }

--- a/elements/userlevel/tonetmapdevice.cc
+++ b/elements/userlevel/tonetmapdevice.cc
@@ -24,8 +24,9 @@
 #include <click/error.hh>
 #include <click/standard/scheduleinfo.hh>
 #include <click/packet_anno.hh>
-#include <vector>
 #include <click/master.hh>
+#include <click/netmapdevice.hh>
+#include <vector>
 #include <sstream>
 
 CLICK_DECLS
@@ -114,7 +115,7 @@ int ToNetmapDevice::initialize(ErrorHandler *errh)
 	if (input_is_pull(0)) {
 		if (_pull_use_select) {
 			for (int i = 0; i < n_queues; i++) {
-				master()->thread(thread_for_queue(i))->select_set().add_select(_device->nmds[i]->fd,this,SELECT_WRITE);
+				master()->thread(thread_for_queue_offset(i))->select_set().add_select(_device->nmds[i]->fd,this,SELECT_WRITE);
 			}
 		}
 
@@ -134,7 +135,7 @@ int ToNetmapDevice::initialize(ErrorHandler *errh)
 		    _iodone[i] = false;
 		    _zctimers[i] = new Timer(this);
 		    _zctimers[i]->initialize(this,false);
-		    _zctimers[i]->move_thread(thread_for_queue(i));
+		    _zctimers[i]->move_thread(thread_for_queue_offset(i));
 	     }
 	}
 
@@ -197,16 +198,16 @@ ToNetmapDevice::push_batch(int port, PacketBatch *b_head)
 #endif
 	if (s.q != NULL) {
 		if (s.q_size < _internal_tx_queue_size) {
-			s.q->prev()->set_next(b_head);
+			s.q->prev()->set_next(b_head->first());
 			s.q_size += b_head->count();
-			s.q->set_prev(b_head->prev());
+			s.q->set_prev(b_head->first()->prev());
 		} else {
 		    //We don't have space to store these packets, but we'll try to send the awaiting packets before dropping them
 		    should_be_dropped = true;
 		}
 
 	} else {
-		s.q = b_head;
+		s.q = b_head->first();
 		s.q_size = b_head->count();
 
 		//If we are not struggling, ask a sync at the end of the batch
@@ -233,9 +234,9 @@ do_send_batch:
     } else {
         if (should_be_dropped) {
             if (s.q_size < _internal_tx_queue_size) {
-                s.q->prev()->set_next(b_head);
+                s.q->prev()->set_next(b_head->first());
                 s.q_size += b_head->count();
-                s.q->set_prev(b_head->prev());
+                s.q->set_prev(b_head->first()->prev());
             } else { //Still not enough place, drop !
 		if (_blocking) {
             		allow_txsync();
@@ -341,8 +342,8 @@ static inline int _send_packet(WritablePacket* p, struct netmap_ring* txring, st
 				if (p->headroom() > 0) {
 					complaint++;
 					if (complaint < 5)
-						click_chatter("Shifting data in ToNetmapDevice. You should avoid this case !");
-					p = static_cast<PacketBatch*>(p->shift_data(-p->headroom())); //Is it better to shift or to copy like if it was not a netmap buffer?
+						click_chatter("Shifting data in ToNetmapDevice. You should avoid sending non-netmap Packets !");
+					p = static_cast<WritablePacket*>(p->shift_data(-p->headroom())); //Is it better to shift or to copy like if it was not a netmap buffer?
 				}
 #  if HAVE_NETMAP_PACKET_POOL //We must return the netmap buffer to the packet pool
 				uint32_t buf_idx = NETMAP_BUF_IDX(txring,p->buffer());
@@ -504,11 +505,11 @@ ToNetmapDevice::run_task(Task* task)
 			PacketBatch* new_batch = input_pull_batch(0,_burst - batch_size);
 			if (new_batch) {
 				if (batch) {
-					batch->prev()->set_next(new_batch);
+					batch->prev()->set_next(new_batch->first());
 					batch_size += new_batch->count();
 					batch->set_prev(new_batch->tail());
 				} else {
-					batch = new_batch;
+					batch = new_batch->first();
 					batch_size = new_batch->count();
 				}
 			}

--- a/elements/userlevel/tonetmapdevice.hh
+++ b/elements/userlevel/tonetmapdevice.hh
@@ -7,12 +7,13 @@
 #include <click/timer.hh>
 #include <click/notifier.hh>
 #include <click/etheraddress.hh>
-#include <click/netmapdevice.hh>
 #include <click/multithread.hh>
 #include <vector>
 #include "queuedevice.hh"
 
 CLICK_DECLS
+
+class NetmapDevice;
 
 /*
  * =title ToNetmapDevice

--- a/include/click/netmapdevice.hh
+++ b/include/click/netmapdevice.hh
@@ -23,12 +23,15 @@
 #define CLICK_NETMAPDEVICE_HH
 
 #if HAVE_NETMAP && CLICK_USERLEVEL
-
 #include <net/if.h>
+#define __UAPI_DEF_IF_NET_DEVICE_FLAGS 0
+#include <linux/if.h>
+
+extern "C" {
 #include <net/netmap.h>
 #define NETMAP_WITH_LIBS 1
 #include <net/netmap_user.h>
-
+}
 #define NS_NOFREE 0x80 //We use this to set that a buffer is shared and should not be freed
 
 

--- a/include/click/packet.hh
+++ b/include/click/packet.hh
@@ -1004,6 +1004,7 @@ class WritablePacket : public Packet { public:
 #endif
 
 # if HAVE_CLICK_PACKET_POOL
+    static PacketPool& get_local_packet_pool();
     static void initialize_local_packet_pool();
 # endif
 
@@ -3141,57 +3142,7 @@ WritablePacket::rewrite_seq(tcp_seq_t seq, const int shift) {
     click_update_in_cksum(&this->tcp_header()->th_sum, t_old_hw, t_new_hw);
 }
 
-
-#if !CLICK_PACKET_USE_DPDK
-/**
- * Release a buffer that is directly allocated, ie not one that use destructor
- */
-inline void
-Packet::release_buffer(unsigned char* head) {
-
-# if HAVE_NETMAP_PACKET_POOL
-                if (NetmapBufQ::is_valid_netmap_buffer(head))
-                    NetmapBufQ::local_pool()->insert_p(head);
-                else
-# endif
-                {
-#  if HAVE_DPDK
-                if (dpdk_enabled)
-                    rte_free(head);
-                else
-#  endif
-                    ::operator delete[](head);
-                }
-}
-
-inline void
-Packet::delete_buffer(unsigned char* head, unsigned char* end
-#if CLICK_BSDMODULE
-        ,unsigned char* m
-#endif
-        ) {
-# ifndef CLICK_NOINDIRECT
-    if (_data_packet) {
-	    _data_packet->kill();
-    }
-# else
-  if (false) {}
-# endif
-# if CLICK_USERLEVEL || CLICK_MINIOS
-    else if (head && _destructor) {
-        if (_destructor != empty_destructor)
-            _destructor(head, end - head, _destructor_argument);
-    } else
-        release_buffer(head);
-# elif CLICK_BSDMODULE
-    if (m)
-	    m_freem(m);
-# endif
-}
-#endif
-
 typedef Packet::PacketType PacketType;
-
 
 inline unsigned char* WritablePacket::getPacketContent()
 {

--- a/lib/flow.cc
+++ b/lib/flow.cc
@@ -327,7 +327,7 @@ FlowClassificationTable::Rule FlowClassificationTable::parse(Element* owner, Str
                     }
 #endif
 
-                    if (maskv & valuev == 0)
+                    if ((maskv & valuev) == 0)
                         dynamic = true;
                 }
                 FlowNodeDefinition* node = new FlowNodeDefinition(owner);

--- a/lib/netmapdevice.cc
+++ b/lib/netmapdevice.cc
@@ -201,9 +201,9 @@ NetmapDevice::~NetmapDevice() {
 WritablePacket* NetmapDevice::make_netmap(unsigned char* data, struct netmap_ring* rxring, struct netmap_slot* slot) {
     WritablePacket* p = WritablePacket::pool_data_allocate();
     if (!p) return NULL;
-    p->initialize();
+    p->initialize(true);
     slot->buf_idx = NETMAP_BUF_IDX(rxring, p->buffer());
-    p->set_buffer(data,rxring->nr_buf_size,slot->len);
+    p->init_buffer(data,rxring->nr_buf_size,slot->len);
     slot->flags = NS_BUF_CHANGED;
     return p;
 }
@@ -215,10 +215,11 @@ PacketBatch* NetmapDevice::make_netmap_batch(unsigned int n, struct netmap_ring*
     if (n <= 0) return NULL;
     struct netmap_slot* slot;
     WritablePacket* last = 0;
+    const bool clear = true;
 
     Timestamp ts = Timestamp::make_usec(rxring->ts.tv_sec, rxring->ts.tv_usec);
 
-    PacketPool& packet_pool = *WritablePacket::make_local_packet_pool();
+    PacketPool& packet_pool = WritablePacket::get_local_packet_pool();
 
     WritablePacket*& _head = packet_pool.pd;
     unsigned int & _count = packet_pool.pdcount;
@@ -231,7 +232,7 @@ PacketBatch* NetmapDevice::make_netmap_batch(unsigned int n, struct netmap_ring*
     //Next is the current packet in the batch
     Packet* next = _head;
     //P_batch_saved is the saved head of the batch.
-    PacketBatch* p_batch = static_cast<PacketBatch*>(_head);
+    PacketBatch* p_batch = PacketBatch::start_head(_head);
 
     int toreceive = n;
     while (toreceive > 0) {
@@ -248,9 +249,9 @@ PacketBatch* NetmapDevice::make_netmap_batch(unsigned int n, struct netmap_ring*
         slot->flags = NS_BUF_CHANGED;
 
         next = last->next(); //Correct only if count != 0
-        last->initialize();
+        last->initialize(clear);
 
-        last->set_buffer(data,NetmapBufQ::buffer_size(),slot->len);
+        last->init_buffer(data,NetmapBufQ::buffer_size(),slot->len);
         cur = nm_ring_next(rxring, cur);
         toreceive--;
         _count --;

--- a/lib/packet.cc
+++ b/lib/packet.cc
@@ -230,6 +230,56 @@ Packet::~Packet()
 #endif
 }
 
+#if !CLICK_PACKET_USE_DPDK
+/**
+ * Release a buffer that is directly allocated, ie not one that use destructor
+ */
+inline void
+Packet::release_buffer(unsigned char* head) {
+
+# if HAVE_NETMAP_PACKET_POOL
+                if (NetmapBufQ::is_valid_netmap_buffer(head))
+                    NetmapBufQ::local_pool()->insert_p(head);
+                else
+# endif
+                {
+#  if HAVE_DPDK
+                if (dpdk_enabled)
+                    rte_free(head);
+                else
+#  endif
+                    ::operator delete[](head);
+                }
+}
+
+inline void
+Packet::delete_buffer(unsigned char* head, unsigned char* end
+#if CLICK_BSDMODULE
+        ,unsigned char* m
+#endif
+        ) {
+# ifndef CLICK_NOINDIRECT
+    if (_data_packet) {
+	    _data_packet->kill();
+    }
+# else
+  if (false) {}
+# endif
+# if CLICK_USERLEVEL || CLICK_MINIOS
+    else if (head && _destructor) {
+        if (_destructor != empty_destructor)
+            _destructor(head, end - head, _destructor_argument);
+    } else
+        release_buffer(head);
+# elif CLICK_BSDMODULE
+    if (m)
+	    m_freem(m);
+# endif
+}
+#endif
+
+
+
 #if !CLICK_LINUXMODULE && !CLICK_PACKET_USE_DPDK
 
 # if HAVE_CLICK_PACKET_POOL
@@ -284,13 +334,22 @@ static PacketPool global_packet_pool = {0,0,0,0};
 
 /** @brief Return the local packet pool for this thread.
     @pre initialize_local_packet_pool() has succeeded on this thread. */
-static CLICK_ALWAYS_INLINE inline PacketPool& local_packet_pool() {
+static CLICK_ALWAYS_INLINE inline
+PacketPool& local_packet_pool() {
 #  if HAVE_MULTITHREAD
     return *thread_packet_pool;
 #  else
     // If not multithreaded, there is only one packet pool.
     return global_packet_pool;
 #  endif
+}
+
+/**
+ * Exportable version of the above
+ */
+PacketPool&
+WritablePacket::get_local_packet_pool() {
+    return local_packet_pool();
 }
 
 /** @brief Create and return a local packet pool for this thread. */

